src/utils/jwt.ts

```diff
@@ -0,0 +1,240 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as validator from './validator';
+import * as jwt from 'jsonwebtoken';
+import { HttpClient, HttpRequestConfig, HttpError } from '../utils/api-request';
+import { Agent } from 'http';
+
+export type Dictionary = {[key: string]: any}
+
+export type DecodedToken = {
+  header: Dictionary;
+  payload: Dictionary;
+}
+
+export interface SignatureVerifier {
+  verify(token: string): Promise<void>;
+}
+
+interface KeyFetcher {
+  fetchPublicKeys(): Promise<{ [key: string]: string }>;
+}
+
+export class UrlKeyFetcher implements KeyFetcher {
+  private publicKeys: { [key: string]: string };
+  private publicKeysExpireAt: number;
+
+  constructor(private clientCertUrl: string, private readonly httpAgent?: Agent) {
+    if (!validator.isURL(clientCertUrl)) {
+      throw new Error(
+        'The provided public client certificate URL is an invalid URL.',
+      );
+    }
+  }
+
+  /**
+   * Fetches the public keys for the Google certs.
+   *
+   * @return A promise fulfilled with public keys for the Google certs.
+   */
+  public fetchPublicKeys(): Promise<{ [key: string]: string }> {
+    if (this.shouldRefresh()) {
+      return this.refresh();
+    }
+    return Promise.resolve(this.publicKeys);
+  }
+
+  private shouldRefresh(): boolean {
+    const publicKeysExist = (typeof this.publicKeys !== 'undefined');
+    const publicKeysExpiredExists = (typeof this.publicKeysExpireAt !== 'undefined');
+    const publicKeysStillValid = (publicKeysExpiredExists && Date.now() < this.publicKeysExpireAt);
+    return !(publicKeysExist && publicKeysStillValid);
+  }
+
+  private refresh(): Promise<{ [key: string]: string }> {
+    const client = new HttpClient();
+    const request: HttpRequestConfig = {
+      method: 'GET',
+      url: this.clientCertUrl,
+      httpAgent: this.httpAgent,
+    };
+    return client.send(request).then((resp) => {
+      if (!resp.isJson() || resp.data.error) {
+        // Treat all non-json messages and messages with an 'error' field as
+        // error responses.
+        throw new HttpError(resp);
+      }
+      if (Object.prototype.hasOwnProperty.call(resp.headers, 'cache-control')) {
+        const cacheControlHeader: string = resp.headers['cache-control'];
+        const parts = cacheControlHeader.split(',');
+        parts.forEach((part) => {
+          const subParts = part.trim().split('=');
+          if (subParts[0] === 'max-age') {
+            const maxAge: number = +subParts[1];
+            this.publicKeysExpireAt = Date.now() + (maxAge * 1000);
```

Comment:
It seems this gets set conditionally. I think for other cases, we should 0 it out just to be on the safe side.

Reply:
Good call! Added the following at line 88 to reset the parameter.
```
// reset expire at from previous set of keys.
this.publicKeysExpireAt = 0;
```